diffkemp/building/cc_wrapper_template.py

```diff
@@ -0,0 +1,213 @@
+
+
+def wrapper(argv):
+    # Get arguments from environmental variables
+    db_filename = os.environ.get(wrapper_env_vars['db_filename'])
+    clang = os.environ.get(wrapper_env_vars['clang'])
+    append = os.environ.get(wrapper_env_vars['clang_append']).split(',')
+    drop = os.environ.get(wrapper_env_vars['clang_drop']).split(',')
+    debug = os.environ.get(wrapper_env_vars['debug'])
+
+    if (db_filename is None or clang is None or append is None or
+            drop is None or argv is None):
+        return 1
+
+    db = []
+    # TODO: Allow custom llvm-link and llvm-dis
+    llvm_link = 'llvm-link'
+    llvm_dis = 'llvm-dis'
+
+    # Run GCC
+    argv[0] = 'gcc'
+    try:
+        check_call('gcc', argv)
+    except CalledProcessError:
+        return 1
+
+    # Analyze and modify parameters for clang (phase 1)
+    clang_argv = []
+    old_clang = clang
+    output_file_flag = False
+    linking_with_sources = False
+    output_file = None
+    linking = '-c' not in argv
+    for arg in argv:
+        if arg in drop:
+            continue
+        is_object_file = (arg.endswith('.o') or arg.endswith('.lo') or
+                          arg.endswith('.ko'))
+        is_source_file = arg.endswith('.c')
+        if arg == '-o':
+            output_file_flag = True
+        elif output_file_flag:
+            output_file_flag = False
+            if is_object_file and not linking:
+                # Compiling to object file: swap .o with .ll
+                arg = arg.rsplit('.', 1)[0] + '.ll'
+            if not is_object_file and linking:
+                # Linking: add a .llw suffix (LLVM IR whole)
+                arg = arg + '.llw'
+            output_file = arg
+        elif is_object_file and linking:
+            # Input to linking phase: change suffix to .ll
+            arg = arg.rsplit('.', 1)[0] + '.ll'
+            clang = llvm_link
+        elif is_source_file and linking:
+            # Mark as linking with sources to detect hybrid mode
+            linking_with_sources = True
+        clang_argv.append(arg)
+
+    if linking_with_sources and clang == llvm_link:
+        # Compile/link mode with object files detected
+        # Drop object files and revert to normal compile/link mode
+        clang = old_clang
+        new_clang_argv = []
+        for arg in clang_argv:
+            if arg.endswith('.ll'):
+                continue
+            new_clang_argv.append(arg)
+        clang_argv = new_clang_argv
+
+    # Do not continue if output is not .ll or .llw
+    # Note: this means that this is neither compilation nor linking
+    if (output_file is not None and not output_file.endswith('.ll') and
+            not output_file.endswith('.llw')):
+        return 0
+
+    # Do not run clang on conftest files
+    if output_file in ['conftest.ll', 'conftest.llw'] or 'conftest.c' in argv:
+        return 0
+
+    # Record file in database
+    if output_file is not None:
+        prefix = 'o:' if clang != llvm_link else 'f:'
```

Comment:
The wrapper only sets `clang` equal to `llvm_link` when object files are being linked, not when all arguments are C sources (both have `linking` set to `True`, since there's no `-c` argument present in either case), so these are not equivalent. 

The wrapper logic readability should be improved though, the naming and the branches are quite confusing to read.